### PR TITLE
Implement foldICmpRemConstant in InstCombineCompares

### LINKED ISSUE #76585: [InstCombine] Missed optimization for `(x * z) % (y * z) == 0` => `x % y == 0`
`(x * z) % (y * z) == 0` => `x % y == 0` when the multiplications/modulus do not overflow:

[alive](https://alive2.llvm.org/ce/z/ylAxft)
```llvm
define dso_local i1 @src1(i8 noundef %0, i8 noundef %1, i8 noundef %2) {
  %4 = mul nuw i8 %2, %0
  %5 = mul nuw i8 %2, %1
  %6 = urem i8 %4, %5
  %7 = icmp eq i8 %6, 0
  ret i1 %7
}

define dso_local i1 @tgt1(i8 noundef %0, i8 noundef %1, i8 noundef %2) {
  %4 = urem i8 %0, %1
  %5 = icmp eq i8 %4, 0
  ret i1 %5
}

define dso_local i1 @src2(i8 noundef %0, i8 noundef %1, i8 noundef %2) {
  %4 = icmp ne i8 %1, -1
  tail call void @llvm.assume(i1 %4)
  %5 = mul nsw i8 %2, %0
  %6 = mul nsw i8 %2, %1
  %7 = srem i8 %5, %6
  %8 = icmp eq i8 %7, 0
  ret i1 %8
}


define dso_local i1 @tgt2(i8 noundef %0, i8 noundef %1, i8 noundef %2) {
  %4 = srem i8 %0, %1
  %5 = icmp eq i8 %4, 0
  ret i1 %5
}

declare void @llvm.assume(i1 noundef)
```

### PR DESCRIPTION
**Pull request by Baxi-codes**
Doesn't work right now

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 1]
```diff
@@ -2572,6 +2572,46 @@ Instruction *InstCombinerImpl::foldICmpSRemConstant(ICmpInst &Cmp,
   return new ICmpInst(ICmpInst::ICMP_UGT, And, ConstantInt::get(Ty, SignMask));
 }
 
+Instruction *InstCombinerImpl::foldICmpRemConstant(ICmpInst &Cmp,
+                                                   BinaryOperator *Rem,
+                                                   const APInt &C) {
+  const ICmpInst::Predicate Pred = Cmp.getPredicate();
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:2578]:
```suggestion
  assert((Rem->getOpcode() == Instruction::SRem || Rem->getOpcode() == Instruction::URem) && "Only for srem/urem!");
  const ICmpInst::Predicate Pred = Cmp.getPredicate();
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 2]
```diff
@@ -2572,6 +2572,46 @@ Instruction *InstCombinerImpl::foldICmpSRemConstant(ICmpInst &Cmp,
...
+  Value *X = Rem->getOperand(0);
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:2579]:
```suggestion
  if (!ICmpInst::isEquality(Pred) || !C.isZero())
    return nullptr;
  Value *X = Rem->getOperand(0);
```
It only holds for `==/!= 0`.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 3]
```diff
@@ -2572,6 +2572,46 @@ Instruction *InstCombinerImpl::foldICmpSRemConstant(ICmpInst &Cmp,
...
+  Value *A = MulX->getOperand(0);
+  Value *C1 = MulX->getOperand(1);
+  Value *B = MulY->getOperand(0);
+  Value *C2 = MulY->getOperand(1);
+
+  const APInt *C1Value, *C2Value;
+
+  // Check if the constants satisfy the condition c1 % c2 == 0.
+  if (!match(C1, m_APInt(C1Value)) || !match(C2, m_APInt(C2Value)) || C1Value->urem(*C2Value) != 0)
+    return nullptr;
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:2603]:
```suggestion
  Value *A, *B;
  const APInt *C1, *C2;
  if (Rem->getOpcode() == Instruction::SRem) {
    if (!match(X, m_NSWMul(m_Value(A), m_APInt(C1))))
      retur nullptr;
    if (!match(Y, m_NSWMul(m_Value(B), m_APInt(C2))))
      retur nullptr;
    if (!C1->srem(*C2).isZero())
      return nullptr;
  }
  else {
     ... // match with m_NUWMul and use urem
  }
```
Be careful of `nsw/nuw` flags.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 4]
```diff
@@ -2572,6 +2572,46 @@ Instruction *InstCombinerImpl::foldICmpSRemConstant(ICmpInst &Cmp,
...
+
+  // Compute the new constant k = c1 / c2.
+  APInt K = C1Value->udiv(*C2Value);
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:2606]: For srem, use `sdiv` instead.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 5]
```diff
@@ -2572,6 +2572,46 @@ Instruction *InstCombinerImpl::foldICmpSRemConstant(ICmpInst &Cmp,
...
+  Type *Ty = Rem->getType();
+
+  // Create a new remainder instruction (a * k) % b.
+  Value *NewRem = Builder.CreateURem(Builder.CreateMul(A, ConstantInt::get(A->getType(), K)), B);
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:2610]: Same problem as above.
BTW, I think we can preserve the NSW/NUW flags for the mul.

**Review by dtcxzyw — CHANGES_REQUESTED** [review]
I applied your patch locally. It works for me:
```
> bin/opt -passes=instcombine test.ll -S
; ModuleID = 'test.ll'
source_filename = "test.ll"

define i1 @src(i8 noundef %0, i8 noundef %1) {
  %3 = mul i8 %0, 3
  %4 = urem i8 %3, %1
  %5 = icmp eq i8 %4, 0
  ret i1 %5
}
```

**Comment by dtcxzyw** [comment]
> Doesn't work right now

Please update your PR description and attach alive2 proofs.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 6]
```diff
@@ -3708,7 +3748,12 @@ Instruction *InstCombinerImpl::foldICmpBinOpWithConstant(ICmpInst &Cmp,
   case Instruction::SRem:
     if (Instruction *I = foldICmpSRemConstant(Cmp, BO, C))
       return I;
+   [[fallthrough]]; 
+  case Instruction::URem:
```
[the remark this answers is not in the document]
Comment (Baxi-codes) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3752]: How do I do this? Do I undo the changes, add tests, and commit, then do the changes again?

[the remark this answers is not in the document]
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3752]: Yes.

**Comment by Baxi-codes** [comment]
Closed this pr, will create a new one

**Comment by Baxi-codes** [comment]
> > Doesn't work right now
> 
> Please update your PR description and attach alive2 proofs.

nsw with srem doesn't verify
https://alive2.llvm.org/ce/z/n7wYL5

**Comment by dtcxzyw** [comment]
> > > Doesn't work right now
> > 
> > 
> > Please update your PR description and attach alive2 proofs.
> 
> nsw with srem doesn't verify https://alive2.llvm.org/ce/z/n7wYL5

https://alive2.llvm.org/ce/z/R73iWv

`sdiv INT_MIN, -1` will overflow.

**Comment by Baxi-codes** [comment]
> > > > Doesn't work right now
> > > 
> > > 
> > > Please update your PR description and attach alive2 proofs.
> > 
> > 
> > nsw with srem doesn't verify https://alive2.llvm.org/ce/z/n7wYL5
> 
> https://alive2.llvm.org/ce/z/R73iWv
> 
> `sdiv INT_MIN, -1` will overflow.

I am not sure how to implement that transformation.
The original transformation turns out to be correct if both nsw and nuw flags are set
https://alive2.llvm.org/ce/z/THjHEk

**Comment by Baxi-codes** [comment]
Please review the new PR
https://github.com/llvm/llvm-project/pull/79383

**Pull request closed without merging**